dev/NavigationView/NavigationView.cpp

```diff
@@ -1594,6 +1734,222 @@ void NavigationView::UpdateVisualStateForDisplayModeGroup(const winrt::Navigatio
+            {
+                return m_topNavRepeater.get().TryGetElement(0);
+            }
+        }
+        return m_leftNavRepeater.get().TryGetElement(0);
+    }();
+
+    if (auto controlFirst = firstElement.try_as<winrt::Control>())
+    {
+        controlFirst.Focus(winrt::FocusState::Keyboard);
+    }
+}
+
+void NavigationView::KeyboardFocusLastItemFromItem(const winrt::NavigationViewItemBase& nvib)
+{
+    auto const ir = [this, nvib]()
+    {
+        if (IsTopNavigationView())
+        {
+            if (IsContainerInOverflow(nvib))
+            {
+                return m_topNavRepeaterOverflowView.get();
+            }
+            else
+            {
+                return m_topNavRepeater.get();
+            }
+        }
+        else
+        {
+            return m_leftNavRepeater.get();
+        }
+    }();
+
+    if (auto itemsSourceView = ir.ItemsSourceView())
+    {
+        auto lastIndex = itemsSourceView.Count() - 1;
+        if (auto lastElement = ir.TryGetElement(lastIndex))
+        {
+            if (auto controlLast = lastElement.try_as<winrt::Control>())
+            {
+                controlLast.Focus(winrt::FocusState::Keyboard);
+            }
+        }
+    }
+}
+
+void NavigationView::RepeaterGettingFocus(const winrt::IInspectable& sender, const winrt::GettingFocusEventArgs& args)
+{
+    auto const inputDevice = args.InputDevice();
+    if (inputDevice == winrt::FocusInputDeviceKind::Keyboard)
+    {
+        if (auto const oldFocusedElement = args.OldFocusedElement())
+        {
+            auto const oldElementParent = winrt::VisualTreeHelper::GetParent(oldFocusedElement);
+            auto const rootRepeater = [this]()
+            {
+                if (IsTopNavigationView())
+                {
+                    return m_topNavRepeater.get();
+                }
+                return m_leftNavRepeater.get();
+            }();
+            // If focus is coming from outside the root repeater, put focus on last focused item
+            if (rootRepeater != oldElementParent)
+            {
+                if (auto const lastFocusedNvi = rootRepeater.TryGetElement(m_indexOflastFocusedItem))
+                {
+                    if (auto const argsAsIGettingFocusEventArgs2 = args.try_as<winrt::IGettingFocusEventArgs2>())
+                    {
+                        if (argsAsIGettingFocusEventArgs2.TrySetNewFocusedElement(lastFocusedNvi))
+                        {
+                            args.Handled(true);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemOnGotFocus(const winrt::IInspectable& sender, winrt::RoutedEventArgs const& e)
+{
+    if (auto nvi = sender.try_as<winrt::NavigationViewItem>())
+    {
+        // Store index of last focused item in order to get proper focus behavior.
+        // No need to keep track of last focused item if the item is in the overflow menu.
```

Comment:
>No need to keep track of last focused item if the item is in the overflow menu [](start = 11, length = 78)

Doesn't this mean we will have a stale value held in m_indexOflastFocusedItem?

Reply:
Not really stale. Since we are saving focus state only for items in the primary list, it just means that if you navigate around in the overflow menu and then dismiss the menu, it will still allow you to go back to the last focused item in the primary list (which is how it works in the current implementation).